smslibrary/src/test/java/com/eis/smslibrary/SMSHandlerTest.java

```diff
@@ -0,0 +1,72 @@
+package com.eis.smslibrary;
+
+import android.content.Context;
+import android.content.IntentFilter;
+import android.telephony.SmsManager;
+
+import com.eis.smslibrary.listeners.SMSSentListener;
+
+import org.junit.Assert;
+import org.junit.Before;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.mockito.Mock;
+import org.mockito.Mockito;
+import org.mockito.runners.MockitoJUnitRunner;
+
+import static org.junit.Assert.assertEquals;
+import static org.mockito.Mockito.when;
+
+@RunWith(MockitoJUnitRunner.class)
+public class SMSHandlerTest {
+
+    @Mock SmsManager managerMock;
+    @Mock Context contextMock;
+
+    private final String PEER_TEXT = "+393423541601";
+    private final String MESSAGE_TEXT = "valid text";
+    private final String SENT_TEXT = (char)0x02 + MESSAGE_TEXT;
+    private final SMSPeer VALID_PEER = new SMSPeer(PEER_TEXT);
+    private final SMSMessage VALID_MESSAGE = new SMSMessage(VALID_PEER, MESSAGE_TEXT);
+    private SMSHandler instance;
+
+    //"exception" class thrown when a message is correctly sent, used with mocks
+    private class MessageCorrectlySent extends RuntimeException{ }
+
+    @Before
+    public void setup(){
+        instance = SMSHandler.getInstance();
+        instance.clearSetup();
+    }
+
+    @Test(expected = IllegalStateException.class)
+    public void sendMessageWithNullContext_throws(){
+        instance.sendMessage(VALID_MESSAGE);
+    }
+
+    @Test
+    public void singletonInstance(){
+        assertEquals(SMSHandler.getInstance(), SMSHandler.getInstance());
+    }
+
+    @Test(expected = NullPointerException.class)
+    public void setupNullContext_throws(){
+        instance.setup(null);
+    }
+
+    @Test(expected = MessageCorrectlySent.class)
+    public void validMessage_isSent(){
+        Mockito.doThrow(new MessageCorrectlySent())
+                .when(managerMock)
+                .sendTextMessage(PEER_TEXT, null, SENT_TEXT, null, null);
```

Comment:
Nice workaround, but it's actually not needed, there is already a mockito method for that.

If you want to check a method invocation on a mock, you can use `verify(...)`, it basically fails the test if the method is not called.
So in this case, you can just put: `verify(managerMock).sendTextMessage(PEER_TEXT, null, SENT_TEXT, null, null);` under line 64.

The `verify(...)` method is very useful in many other situations, for example, checking that a method is called n times, here's a handy [cookbook](https://www.baeldung.com/mockito-verify).

Reply:
Wow, what a beautiful way to do it. Thanks a lot